### ISSUE TITLE
`DepthLoss26`: the gradient-pyramid sparsity guard is batch-global, so each image's supervision depends on the other images in the batch

### ISSUE THREAD

**Issue by JESUSROYETH** [ultralytics/ultralytics#25615]
Hi 👋

While reading `DepthLoss26` I stumbled on a batch coupling in the multi-scale gradient loss. The sparsity guard that decides whether to descend the pyramid checks the mean valid fraction of the whole batch (`if vp.mean() < 0.5: break`, https://github.com/ultralytics/ultralytics/blob/4e21eac2b0c83f5f2ac75979d5d94fcee209aaf1/ultralytics/utils/loss.py#L1247-L1249), but GT sparsity is a per-image property. So the same image gets different gradient supervision depending on which images share the batch, in both directions:

1. A dense image loses the 3 coarse pyramid levels when its companions drag the batch mean below 0.5: the gradient of the grad term on its prediction falls 2.35x. This is measured at constant batch size (B=3) with identical predictions in both runs, so the only change between the two runs is the companions' GT.
2. The other way round: a sparse image (5 % valid, LiDAR-like scatter) batched with dense companions passes the guard and receives the 3 coarse levels — the gradients the guard is supposed to block. At constant B=4 with zero-error companions, opening the pyramid multiplies the dgrad gradient on the sparse image by ~130x in this repro (the factor depends strongly on how the invalid pixels are distributed).

Environment: Ultralytics 8.4.115 (`main` @ 4e21eac), Python 3.12.3, torch 2.12.1+cu130, CPU. Minimal repro for case 1, defaults from `default.yaml` with `dlog=0` to isolate the gradient term:

```python
import torch, torch.nn as nn, torch.nn.functional as F, types
from ultralytics.utils.loss import DepthLoss26

m = nn.Linear(1, 1)
m.args = types.SimpleNamespace(dlog=0.0, dgrad=0.5, dlam=1.0)
crit = DepthLoss26(m)

g = torch.Generator().manual_seed(42)
gt = F.avg_pool2d(F.pad(torch.rand((1, 1, 64, 64), generator=g) * 9 + 1, (2, 2, 2, 2), mode="replicate"), 5, stride=1)
pred = gt * 1.3 + 0.4 * torch.rand((1, 1, 64, 64), generator=g)
gt_comp = F.avg_pool2d(F.pad(torch.rand((2, 1, 64, 64), generator=g) * 9 + 1, (2, 2, 2, 2), mode="replicate"), 5, stride=1)
pred_comp = gt_comp.clone()  # companion predictions, identical in both runs

for name, gt_c in [
    ("companions valid  (pyramid runs)", gt_comp),
    ("companions invalid (pyramid cut)", torch.zeros((2, 1, 64, 64))),
]:
    p = torch.cat([pred, pred_comp]).clone().requires_grad_(True)
    t = torch.cat([gt, gt_c])
    total, items = crit({"depth": p}, {"depth": t.squeeze(1)})
    total.sum().backward()
    print(f"{name}: dgrad_loss={items['dgrad_loss']:.4f}  ||grad on the dense image||={p.grad[0].norm():.5f}")
```

```
companions valid  (pyramid runs): dgrad_loss=0.0120  ||grad on the dense image||=0.00725
companions invalid (pyramid cut): dgrad_loss=0.0062  ||grad on the dense image||=0.00308
```

(The `dgrad_loss` item and the gradient norm move by different factors — a loss value and the norm of its derivative need not scale together. The gradient is the quantity that trains, so the 2.35x is the relevant readout.)

Does this fire on real data? Measuring the official DIODE val masks (`depth-diode.yaml` ships indoor + outdoor in one split, and zeroes invalid pixels with the official mask at conversion): indoors n=325, median valid fraction 0.998, 0 % of images below 0.5; outdoor n=446, median 0.857, 11.7 % below 0.5 (min 0.114). Two real outdoor masks (0.114 and 0.164) next to one dense image already give a batch mean of 0.426, so at small batch sizes case 1 fires with real DIODE data. At batch=16 the opposite happens: simulating shuffled batches over the val mix (pre-augmentation), the batch mean never dropped below 0.5 in 2000 draws, so the guard never fires — while 67 % of the batches contain at least one sub-0.5 image that a per-image version of the same threshold would gate. And with multi-source training (the `DepthTrainer` docstring mentions passing a list of sources), e.g. KITTI plus a dense dataset, both directions can occur depending on the mix, the sampler and the batch size — I did not train a real multi-source run.

I think the batch-global check survived because the recorded validations (the #25171 table: "guard == single-scale on KITTI exactly, dense unchanged") were aggregate results of KITTI-only and dense-only runs — invariance to batch composition was never tested. To be clear on scope: what I measured is the coupling of the supervision signal, I did not train models to show an accuracy regression.

I did not send a PR directly because the fix is less obvious than it looks. Gating per image on the plain valid fraction also changes the dense path: with the default augmentation (`scale=0.5`), roughly a third of the canvases coming from fully dense GT fall below 0.5 valid just from zoom-out padding (30 ± 2 % across seeded runs of the real train pipeline on `depth8`), and contiguous padding is exactly the case the masked pooling already handles well, it is not LiDAR scatter. So a per-image guard probably wants to look at how scattered the invalid pixels are (for example the fraction of partially-valid pooled cells), not at the raw valid fraction. That choice deserves your experiment suite rather than my synthetic benches.

I'm happy to open a PR with whichever per-image semantics you prefer .. let me know what you think.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JESUSROYETH, thank you for the detailed report and reproducible analysis 🚀! We appreciate the clear explanation of the potential batch-coupling behavior in `DepthLoss26`, including the provided MRE and real-data observations.

This is an automated response 🤖, and an Ultralytics engineer will also assist with reviewing the reported behavior soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples, along with answers to many common questions.

Since this is a 🐛 Bug Report, the provided [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) is very helpful. Please also include any additional training logs, dataset details, or experiment results that may help validate the behavior across different batch compositions. For custom training, please verify that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package, including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify whether the behavior is already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments, with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks, this confirms an unintended batch-composition dependency. Please open a minimal PR using a per-image continuation mask based on pooled-cell occupancy/partial-cell scatter rather than raw valid fraction, with tests covering batch-composition invariance, dense masks with augmentation padding, and sparse LiDAR-like masks; we can tune the exact threshold from those results.

**Comment by Daniiiil1** [comment]
Hi @glenn-jocher — I would like to implement the requested minimal PR. I will begin with a batch-composition invariance test, then apply a per-image continuation mask based on pooled-cell occupancy, covering dense masks with augmentation padding and sparse LiDAR-like masks. Should I target current `main`, and is `1e-6` an appropriate gradient-invariance tolerance for the regression test?

— Daniil Mordanov

**Comment by glenn-jocher** [comment]
Please coordinate on the existing [PR #25637](https://github.com/ultralytics/ultralytics/pull/25637) rather than opening a duplicate. Any follow-up should target current `main`; for cross-device gradient checks, prefer `torch.testing.assert_close(..., rtol=1e-5, atol=1e-8)` over a single `1e-6` tolerance.
